### ISSUE TITLE
LOO error when using Student's t SAR model

### ISSUE THREAD

**Issue by aslez** [paul-buerkner/brms#705]
To test the new SAR routines, I ran two models using the following:
data(oldcol, package = "spdep")
brms_lagsar_norm <- brm(CRIME ~ INC + HOVAL, data = COL.OLD, 
                   autocor = cor_lagsar(COL.nb),
                   chains = 2, cores = 2)
brms_lagsar_t <- brm(CRIME ~ INC + HOVAL, data = COL.OLD, family = "student",
                        autocor = cor_lagsar(COL.nb),
                        chains = 2, cores = 2)

loo(brms_lagsar_norm) works, but loo(brms_lagsar_t) gives the following error:

Error: Cannot yet compute pointwise log-likelihood for this model because the observations are not conditionally independent.

The process of doing this for the Student's t model should be similar to the normal model, though I realize that you may not have worked this out explicitly.

Finally, I take back what I said on the earlier thread.  If speed turns out to be issue when using the old SAR parameterization to calculate loo measures, you could replace

g <- Cinv %*% (y - solve(W_tilde, eta))

with something like

g <- Cinv %*% (y - (I_n + p$lagsar * fit$autocor$W + p$lagsar^2 * fit$autocor$W^2) %*% eta).

You would need to figure out how far to extend the series to get a decent approximation (i.e. I don't think the approximation above is close enough).  At the same time, the speed of the loo command with the normal model seems okay to me.  I'll have to see how it does with a larger data set.

**Comment by paul-buerkner** [comment]
Yes indeed. For the normal model we could rely on standard multvariate normal theory to figure out the factorization in terms of p(y_i | y_{-i}, theta), but for the student-t model, we did not dive into that. If you can provide a solution, I am very happy to implement it. The series approximation needs to be further studied before I would consider it an option for brms, but it is definitely something to keep in mind.

**Comment by aslez** [comment]
I'm not sure that I can provides a proper solution, but I'm also wondering whether a formal proof is really necessary here.  We can get to a Student's-t model by placing priors on the variance-covariance matrix of Gaussian model, right?  If that's the case, would anything change in the existing setup beyond the resulting likelihood formula?

I'll see if I can find anything on the series approximation.  The quality of the approximation may vary too much depending on the structure of the weights matrix and the degree of autocorrelation to be reliable.

**Comment by paul-buerkner** [comment]
Perhaps you are right with regard to the student-t model. To be clear, I am thinking of the following scenario:

A normal SAR model can be written as y ~ multi_normal(mu, Sigma) with mu and Sigma appropriately chosen. This implies a conditional distibution of y_i ~ normal(mu_i, sigma_i) with mu_i and sigma_i depending on mu, Sigma, and y_{-i}. Because y is (multivariate) normal, we know that y_i is normal as well. 

The question is, how does that generalize to student-t SAR models? Here we have y ~ multi_student_t(nu, mu, Sigma), but what does that imply for the distribution for y_i. I am not sure
that we can simply write y_i ~ student_t(nu, mu_i, sigma_i) or is this trivial to see?

**Comment by aslez** [comment]
This reviews relationship between Gaussian- and t-processes: https://arxiv.org/pdf/1402.4306.pdf  

I believe the  relevant conditionals (or something close to them) are laid out in Section 4.2.

**Comment by paul-buerkner** [comment]
Yes, equation (6) in Section 4.2 is exactly what we need. Thank you! I will implement this approach in the next couple of days.

**Comment by aslez** [comment]
Fantastic!

**Comment by paul-buerkner** [comment]
I have to take that back partially. The provided equations provide a
solution but require to compute the inverse of an (n-1) × (n-1) covariance
matrix s × n times where s is the number of posterior samples and n is the
number of observations. This quickly becomes computationally unfeasible.
The elegance of the equations we use for the gaussian case is that we need
the inversion pf a n × n matrix only s times (instead of s × n).

What we need is thus to transform the student t case in the same way but I
haven't gotten my head around yet to solve this.

aslez <notifications@github.com> schrieb am Sa., 13. Juli 2019, 22:53:

**Comment by aslez** [comment]
Could you use something like the Sherman-Morrison-Woodberry approach?

**Comment by paul-buerkner** [comment]
Indeed, I think this is more or less what Sundararajan & Keerthi (2001) are using in their paper "Predictive Approaches for Choosing Hyperparameters in Gaussian Processes", which provides the basis for the approach we advocate in our non-factorizable paper. The proof of the efficient equations for the conditional means and sds is in their appendix. (If you can't access the paper version with the appendix, please email me and I will send it to you). 

Now, we have to think of how to move this to the student-t case. Basically, what we need are efficient expressions of $\tilde{phi}_2$,  $K_{22}$, and $\beta_1$ from equation (6) in Shah et al. if the subset 2 consists of just a single observation. For $\tilde{phi}_2$,  $K_{22}$ we have those already as they are identical to the gaussian case, so that the efficient computation of $\beta_1$ for each left-out observation remains the only problem. That is, we should express $\beta_1$ in terms of the (inverse of) the full covariance matrix K rather then the (inverse of) the subset covariance matrix $K_{22}$ obtained by each time leaving out a single observation.

The proof of that may actually be very similar to the proof of the other quantities, but I didn't have time to look at this in detail myself. If you want to take a look you are more than welcome!

**Comment by aslez** [comment]
As a point of clarification, do you mean that we want $\beta_1$ in terms of the (inverse of) $K$ instead of the (inverse of) the subset covariance matrix $K_{11}$?  If so, couldn't you just solve for the (inverse of) the subset covariance matrix $K_{11}$ based on the formula for $\tilde{phi}_2$?  Other than the (inverse of) $K_{11}$, everything in the formula for $\tilde{phi}_2$ should be a vector when $n_2$ = 1.  Does that seem right to you?  If not right, at least doable?

**Comment by paul-buerkner** [comment]
This is indeed what I mean.

I may be able to "just solve" this in a way you suggested, but I currently
don't have time to check this myself.
If you can provide an expression of beta_1 that works simply on the basis
of (the inverse of) $K_{11}$ rather than
on the basis of its submatrices, I would highly appreciate it.

Am Fr., 19. Juli 2019 um 01:53 Uhr schrieb aslez <notifications@github.com>:

**Comment by aslez** [comment]
I looked into this but ended up deleting the post when I realized (duh) that the vectors couldn’t be inverted the way they needed to be.  I’m working on other stuff at the moment, but will gladly look again.

On Aug 3, 2019, at 3:26 PM, Paul-Christian Bürkner <notifications@github.com<mailto:notifications@github.com>> wrote:

This is indeed what I mean.

I may be able to "just solve" this in a way you suggested, but I currently
don't have time to check this myself.
If you can provide an expression of beta_1 that works simply on the basis
of (the inverse of) $K_{11}$ rather than
on the basis of its submatrices, I would highly appreciate it.

Am Fr., 19. Juli 2019 um 01:53 Uhr schrieb aslez <notifications@github.com<mailto:notifications@github.com>>:

**Comment by paul-buerkner** [comment]
Yes, this is what I also stumbled over when trying the direct path of solving this problem. I think following the proofs for the other quantities, we should be able to obtain an expression for beta_1.
So whenever some of us has the time, I think we should be able to get it working.

**Comment by paul-buerkner** [comment]
It took some months but I finally managed to take a look at the student-t case today. I think I found a proof for non-factorizable student-t models which enables pointwise log-likelihood computations as efficienctly as in the Gaussian case. That is, computing pointwise log-lik values for N observations and S posterior samples now requires O(S * N^k) operations where O(N^k) is the complexity of the matrix inversion, that is, k between 2.x and 3. In comparison, the brute force approach using only basic theory of conditional gaussian/student-t models would have required O(S * N^(k +1)) operations.

I doubt that we can make it more efficient than O(S * N^k) as the covariance matrix inversion is required in any case and needs to be done idependently for each posterior sample S if the covariance matrix depends on unknown parameters. Of course, for covariance matrices which only depend on data, we have O(N^k) as the full complexity.
